sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala

```diff
@@ -103,7 +103,16 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
       typeSoFar
     } else {
       typeSoFar match {
-        case NullType => tryParseInteger(field)
+        case NullType =>
+          tryParseInteger(field) match {
```

Comment:
Let's don't have a separate rule alone here. The problem here seems like when the type is inferred from single partition alone, and the compatible types are not respected. You might have to call `compatibleType` from my cursory look.

Reply:
But the problem is that that's going to degrade the performance, and single partition case is arguably rare. Let's make sure we don't have the performance penalty in general cases.